### PR TITLE
Task sf/j65 add comment functionality

### PR DESCRIPTION
**Pull request by Sayonarko**

### CODE REVIEW

File: backend/src/data/repositories/comment/comment.repository.ts [hunk 1]
```diff
@@ -21,7 +21,10 @@ class Comment {
   }
 
   public getAllByEpisodeId(id: number): Promise<TComment[]> {
-    return this.#CommentModel.query().where('episode_id', id).withGraphFetched('user');
+    return this.#CommentModel.query()
+      .where('episode_id', id)
+      .withGraphFetched('user')
+      .orderBy('created_at', 'desc');
```
Comment (ghost) [inline backend/src/data/repositories/comment/comment.repository.ts]:
```suggestion
```

**Comment by Sayonarko** [comment]
[Add comment functionality](https://trello.com/c/GdHVWSAe/65-add-comment-functionality)

File: shared/src/validation-schemas/comment/comment.validation-schema.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import { Joi } from '~/helpers/helpers';
+import {
+  CommentValidationRule,
+  CommentValidationMessage,
+  CommentCreatePayloadKey,
+} from '~/common/enums/enums';
+import { CommentCreatePayload } from '~/common/types/types';
+
+const comment = Joi.object<CommentCreatePayload>({
```
Comment (ghost) [inline shared/src/validation-schemas/comment/comment.validation-schema.ts:9]: we need two validation

one for the backend
the second for frontend

check the episode validation example

in the shared folder we store only the common things

File: shared/src/common/constants/episode/episode.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+const DEFAULT_EPISODE_ID = -1;
+
+export { DEFAULT_EPISODE_ID };
```
Comment (ghost) [inline shared/src/common/constants/episode/episode.ts]:
```suggestion
```

File: frontend/src/store/episode/actions.ts [hunk 4]
```diff
@@ -18,4 +24,16 @@ const loadCommentsByEpisodeId = createAsyncThunk<Comment[], number, AsyncThunkCo
...
 });
 
-export { loadEpisode, loadCommentsByEpisodeId };
+const createComment = createAsyncThunk<Comment, CommentCreatePayload, AsyncThunkConfig>
+(ActionType.CREATE_COMMENT, async (payload, { extra, getState }) => {
+  const { commentApi } = extra;
+  const { auth } = getState();
+  const comment = await commentApi.create({
+    ...payload,
+    userId: (<User>auth.user).id,
```
Comment (ghost) [inline frontend/src/store/episode/actions.ts:33]: what about episode it?

File: frontend/src/components/episode/episode.tsx [hunk 5]
```diff
@@ -1,54 +1,63 @@
...
+    user: auth.user,
   }));
 
   useEffect(() => {
     dispatch(episodeActions.loadCommentsByEpisodeId(Number(id)));
     dispatch(episodeActions.loadEpisode(Number(id)));
   }, []);
 
+  const handleCreateComment = (payload: CommentCreatePayload): void => {
+    payload.episodeId = Number(id);
```
Comment (ghost) [inline frontend/src/components/episode/episode.tsx]:
```suggestion
```
do it in the action

File: frontend/src/components/episode/components/create-comment-form/create-comment-form.tsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  user: User | null;
+  onSubmit: (payload: CommentCreatePayload) => void;
+};
+
+const CreateCommentForm: React.FC<Props> = ({ user, onSubmit }) => {
+  const { control, handleSubmit, errors } = useAppForm({
+    validationSchema: commentCreateValidationSchema,
+    defaultValues: DEFAULT_CREATE_COMMENT_PAYLOAD,
+  });
+  const isAuth = Boolean(user);
```
Comment (ghost) [inline frontend/src/components/episode/components/create-comment-form/create-comment-form.tsx]:
```suggestion
```

File: frontend/src/components/episode/components/create-comment-form/create-comment-form.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
+import { CommentCreatePayloadKey, ButtonType, AppRoute } from 'common/enums/enums';
+import { CommentCreatePayload, User } from 'common/types/types';
+import { useAppForm } from 'hooks/hooks';
+import { Input, Button, Link } from 'components/common/common';
+import { comment as commentCreateValidationSchema } from 'validation-schemas/validation-schemas';
+import { DEFAULT_CREATE_COMMENT_PAYLOAD } from './common/constants';
+import styles from './styles.module.scss';
+
+type Props = {
+  user: User | null;
```
Comment (ghost) [inline frontend/src/components/episode/components/create-comment-form/create-comment-form.tsx]:
```suggestion
```

File: frontend/src/components/episode/components/create-comment-form/create-comment-form.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+          >
+            {' '}
+            Sign Up
+            {' '}
+          </Link>
+          to write a comment
+        </span>
+      </div>
+    );
+  }
```
Comment (ghost) [inline frontend/src/components/episode/components/create-comment-form/create-comment-form.tsx]:
```suggestion
```

File: frontend/src/components/episode/episode.tsx [hunk 9]
```diff
@@ -1,54 +1,63 @@
...
+          </p>
+        </div>
+      ) : (
+        <h1 className={styles.notFound}>Oops. There is no such episode</h1>
+      )}
       <div className={styles.commentsWrapper}>
+        <CreateCommentForm
+          user={user}
+          onSubmit={handleCreateComment}
+        />
```
Comment (ghost) [inline frontend/src/components/episode/episode.tsx]:
```suggestion
        {hasUser && <CreateCommentForm
          onSubmit={handleCreateComment}
        />
```

File: frontend/src/components/episode/components/create-comment-form/common/constants.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import { CommentCreatePayloadKey } from 'common/enums/enums';
+import { CommentCreatePayload } from 'common/types/types';
+import { DEFAULT_USER_ID, DEFAULT_EPISODE_ID } from 'common/constants/constants';
+
+const DEFAULT_CREATE_COMMENT_PAYLOAD: CommentCreatePayload = {
+  [CommentCreatePayloadKey.TEXT]: '',
+  [CommentCreatePayloadKey.USER_ID]: DEFAULT_USER_ID,
+  [CommentCreatePayloadKey.EPISODE_ID]: DEFAULT_EPISODE_ID,
```
Comment (ghost) [inline frontend/src/components/episode/components/create-comment-form/common/constants.ts]:
```suggestion
```
check example of the episode validation

**Review by ghost — CHANGES_REQUESTED** [review]

File: frontend/src/store/episode/actions.ts [hunk 11]
```diff
@@ -18,4 +24,17 @@ const loadCommentsByEpisodeId = createAsyncThunk<Comment[], number, AsyncThunkCo
...
 
-export { loadEpisode, loadCommentsByEpisodeId };
+const createComment = createAsyncThunk<Comment, CommentFormCreatePayload, AsyncThunkConfig>
+(ActionType.CREATE_COMMENT, async (payload, { extra, getState }) => {
+  const { commentApi } = extra;
+  const { auth, episode: { episode } } = getState();
+  const comment = await commentApi.create({
+    ...payload,
+    userId: (<User>auth.user).id,
+    episodeId: (<Episode>episode).id,
```
Comment (ghost) [inline frontend/src/store/episode/actions.ts]:
```suggestion
  const { auth, episode } = getState();
  const comment = await commentApi.create({
    ...payload,
    userId: (<User>auth.user).id,
    episodeId: (<Episode>episode.episode).id,
```

File: frontend/src/store/episode/reducer.ts [hunk 12]
```diff
@@ -37,6 +37,9 @@ const reducer = createReducer(initialState, (builder) => {
   builder.addCase(loadCommentsByEpisodeId.rejected, (state) => {
     state.dataStatus = DataStatus.REJECTED;
   });
+  builder.addCase(createComment.fulfilled, (state, action) => {
+    state.comments.push(action.payload);
+  });
```
Comment (ghost) [inline frontend/src/store/episode/reducer.ts]: you need to add `commentDataStatus` to the reducer, and disable the form while creating comment

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**